pkg/didcomm/common/service/service.go

```diff
@@ -20,16 +27,61 @@ type DIDComm interface {
 	Event
 }
 
+// Header helper structure which keeps reusable fields
+type Header struct {
+	ID     string           `json:"@id"`
+	Thread decorator.Thread `json:"~thread"`
+	Type   string           `json:"@type,omitempty"`
+}
+
 // DIDCommMsg did comm msg
 type DIDCommMsg struct {
 	// Outbound indicates the direction of this DIDComm message:
 	//   - outgoing (to another agent)
 	//   - incoming (from another agent)
 	Outbound bool
-	Type     string
 	Payload  []byte
 	// TODO : might need refactor as per the issue-226
 	OutboundDestination *Destination
+	Header              *Header
+}
+
+// GetHeader parses and returns Header from payload
+func (m *DIDCommMsg) GetHeader() (*Header, error) {
+	if m.Header != nil {
+		return m.Header, nil
+	}
+
+	if err := json.Unmarshal(m.Payload, &m.Header); err != nil {
```

Comment:
this should be done when DIDComMsg was constructed not in an accessor.

due this being done in an accessor, thread safety concerns are introduced and complexity is increased.

Reply:
Done!